ONSdigital/eq-questionnaire-runner
# Using supplementary data non list items

### What is the context of this PR?
This PR implements the usage of supplementary data (non list items) in a survey. This adds a new option for value sources of `supplementary_data` which has an identifier refering to a top level key of the supplementary data, and an optional list of selectors for accessing nested data

the changes are supported by [the corresponding validator PR](https://github.com/ONSdigital/eq-questionnaire-validator/pull/209)

### Background (read before you review)
To get this working, the only real change, is adding the `SupplementaryDataStore` to `ValueSourceResolver` and adding one additional method to resolve supplementary data. Unfortunately the number of dependencies for `ValueSourceResolver` is very high. There are many different functions and classes which initialise the class, and many functions and classes that in turn depend on those. It ended up being around 93 different functions/classes that needed the store to be added, which get called in about 330 different places

To ease the burden of the developers reviewing this PR, I have split it into 3 distinct commits. 
[Add SupplementaryDataStore dependency](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1169/commits/ece4f5ff0d552fd284c169b76d4d2ed53a4d0051)
The first commit is purely adding the supplementary data store as a parameter everywhere that it needs to go, tests, classes, functions, etc. Functionally no change, tests still pass, 100% coverage

[Using Supplementary data non list items](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1169/commits/419174dfbbfa60bb717c7b573333ce7fa322c1cc)
The second commit is adding the actual functionality to be able to use supplementary data, which is just a change to value source resolver and an adjustment to the test data being used by the mock endpoint, to show it working.

[Testing Supplementary Data non list items](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1169/commits/0c0b42a9aa29ae4dbf0b15a3d315fa475ab1b5af)
The final commit is the tests. The functional and integration tests will only work if the mock endpoint is running. (These will fail in the pipeline until we have replaced this with a dockerised version)

A couple of other commits since for PR comments/Linting
[More tests](https://github.com/ONSdigital/eq-questionnaire-runner/pull/1169/commits/cf7cc35d18526b3b1dcfcefc12825f83f7dc4410)

### How to review
Read through the trello card to understand the requirements and check that the test schema and tests have covered what is asked for.

In terms of reviewing code, I would suggest skim reading the first commit, then checking the other two in depth.

### How to test functionally
Following the guidance in run-mock-sds-endpoint.md start the mock sds api and pick the `test_supplementary_data_non_list_items` schema. 
The dropdown for sds dataset id will show two different ids, these are the same structure but with different data, check that placeholders render correctly with each of them. You should also be able to preserve your response id, relaunch the survey with the updated dataset which has some different data and see your original answers against the new placeholders.

### Checklist
